internals/search_fulltext.py

```diff
@@ -0,0 +1,216 @@
+    feature_id = fe.key.integer_id()
+    feature_words = (existing_fw_dict.get(feature_id) or
+                     FeatureWords(feature_id=feature_id))
+
+    words = parse_words(get_strings(fe))
+    logging.info('feature %r has words %r', feature_id, words)
+    feature_words.words = words
+    updated_fw_list.append(feature_words)
+
+  return updated_fw_list
+
+
+def index_feature(fe):
+  """Create or update a word bag for the given feature entry."""
+  feature_id = fe.key.integer_id()
+  query = FeatureWords.query(FeatureWords.feature_id == feature_id)
+  existing_fw_list = query.fetch(None)
+  updated_fw_list = batch_index_features([fe], existing_fw_list)
+  updated_fw_list[0].put()
+
+
+def search_fulltext(textterm):
+  """Return IDs of features that have some word(s) from phrase."""
+  search_words = parse_words([textterm])
+  if not search_words:
+    logging.warn('Cannot process fulltext term: %r', textterm)
+    return []  # user is searching for stop words.
+
+  logging.info('looking for words: %r', search_words)
+  query = FeatureWords.query(FeatureWords.words.IN(search_words))
+  feature_projections = query.fetch(projection=['feature_id'])
+  feature_ids = [proj.feature_id for proj in feature_projections]
+  return feature_ids
+
+
+# TODO(jrobbins): Implement phrase search as post-procesing.  I.e.,
+# searching for phrase "build useful stuff" would first be handled as
+# a search for any combination of those words.  Then, after all search
+# conditions have been applied, a post-processing step would load all
+# potentially matching features and keep only the ones that contain
+# the entire phrase in one string.
+
+# TODO(jrobbins): Likewise, searching for a word or phrase in a
+# specific field of a feature entry can be done with post-processing.
+
+
+class ReindexAllFeatures(FlaskHandler):
+
+  def get_template_data(self, **kwargs):
+    """Updates the fulltext index for all features."""
+    self.require_cron_header()
+
+    all_feature_entries = FeatureEntry.query().fetch()
+    all_feature_words = FeatureWords.query().fetch()
+    updated_fw_list = batch_index_features(
+        all_feature_entries, all_feature_words)
+    ndb.put_multi(updated_fw_list)
+    msg = f'Added or updated {len(updated_fw_list)} FeatureWords'
+    logging.info(msg)
+    return msg
+
+
+class FindStopWords(FlaskHandler):
+
+  def get_template_data(self, **kwargs):
+    """Count occurances of all words and return the 100 most common."""
+    self.require_cron_header()
+
+    all_feature_entries = FeatureEntry.query().fetch()
+    all_feature_words = []
+    updated_fw_list = batch_index_features(
+        all_feature_entries, all_feature_words)
+    # Don't actually store them, we just want to report stats.
+
+    counts = collections.Counter()
+    for fw in updated_fw_list:
+      counts.update(fw.words)
+
+    logging.info('100 most common words')
+    for word, count in counts.most_common(100):
+      logging.info('%6d: %s', count, word)
```

Comment:
I haven't used a Counter since interview practice 😆 Great time to use it!

Reply:
In my next PR, I'll show how to sort millions of unsigned short ints in O(n) time :).